File: hierarchical-manet.cc
#include "ns3/core-module.h" // Módulo central de NS-3 para configuración básica
#include "ns3/network-module.h" // Permite la creación y manejo de nodos y dispositivos de red
#include "ns3/mobility-module.h" // Define modelos de movilidad para los nodos
#include "ns3/internet-module.h" // Proporciona la pila de protocolos TCP/IP
#include "ns3/wifi-module.h" // Permite la configuración de dispositivos y canales Wi-Fi
#include "ns3/applications-module.h" // Contiene aplicaciones de tráfico como OnOff y FTP
#include "ns3/flow-monitor-module.h" // Herramienta para monitorear y analizar el rendimiento de la red

using namespace ns3; // Espacio de nombres de NS-3

NS_LOG_COMPONENT_DEFINE("HierarchicalMANET"); // Define un componente de log para rastrear la simulación

// Función para configurar un clúster
void ConfigureCluster(Ptr<Node> clusterHead, NodeContainer clusterNodes, double range) {
    // Configurar los dispositivos Wi-Fi
    WifiHelper wifi; // Ayuda para configurar Wi-Fi
    wifi.SetStandard(WIFI_PHY_STANDARD_80211g); // Configura el estándar Wi-Fi 802.11g

    WifiMacHelper mac; // Ayuda para configurar la capa MAC
    mac.SetType("ns3::AdhocWifiMac"); // Configura el modo Ad-Hoc

    YansWifiPhyHelper phy = YansWifiPhyHelper::Default(); // Configuración física Wi-Fi por defecto
    YansWifiChannelHelper channel = YansWifiChannelHelper::Default(); // Configura el canal inalámbrico por defecto
    phy.SetChannel(channel.Create()); // Asocia el canal con el dispositivo físico

    // Instalar dispositivos Wi-Fi en los nodos del clúster
    NetDeviceContainer devices = wifi.Install(phy, mac, clusterNodes);
    devices.Add(wifi.Install(phy, mac, clusterHead)); // Agregar el nodo líder del clúster

    // Instalar la pila de protocolos TCP/IP
    InternetStackHelper internet;
    internet.Install(clusterNodes); // Instala la pila en los nodos del clúster
    internet.Install(clusterHead); // Instala la pila en el líder del clúster

    // Asignar direcciones IP a los nodos
    Ipv4AddressHelper address;
    address.SetBase("10.1.0.0", "255.255.255.0"); // Rango de direcciones IP para el clúster
    address.Assign(devices); // Asigna las direcciones a los dispositivos Wi-Fi

    // Configurar la movilidad de los nodos
    MobilityHelper mobility;
    Ptr<ListPositionAllocator> positionAlloc = CreateObject<ListPositionAllocator>();
    for (uint32_t i = 0; i < clusterNodes.GetN(); ++i) {
        positionAlloc->Add(Vector(range + i * 10.0, range, 0.0)); // Posiciona los nodos en una línea
    }
    positionAlloc->Add(Vector(range, range, 0.0)); // Posiciona el líder del clúster
    mobility.SetPositionAllocator(positionAlloc);
    mobility.SetMobilityModel("ns3::RandomWalk2dMobilityModel", // Configura un modelo de movimiento aleatorio
                              "Bounds", RectangleValue(Rectangle(-range, range, -range, range))); // Define límites de movimiento
    mobility.Install(clusterNodes); // Aplica la movilidad a los nodos del clúster
    mobility.Install(clusterHead); // Aplica la movilidad al líder del clúster
}

int main(int argc, char *argv[]) {
    LogComponentEnable("HierarchicalMANET", LOG_LEVEL_INFO); // Habilita los logs para rastrear la ejecución

    // Crear nodos para los clústeres
    NodeContainer cluster1, cluster2, bridgeCluster;
    cluster1.Create(5); // Clúster 1 con 5 nodos
    cluster2.Create(5); // Clúster 2 con 5 nodos
    bridgeCluster.Create(1); // Nodo puente para conectar los clústeres

    Ptr<Node> cluster1Head = CreateObject<Node>(); // Nodo líder del clúster 1
    Ptr<Node> cluster2Head = CreateObject<Node>(); // Nodo líder del clúster 2

    // Configurar clústeres
    ConfigureCluster(cluster1Head, cluster1, 50.0); // Configura el clúster 1
    ConfigureCluster(cluster2Head, cluster2, 150.0); // Configura el clúster 2

    // Configurar puente entre clústeres
    NodeContainer bridgeNodes;
    bridgeNodes.Add(cluster1Head); // Agregar líder del clúster 1
    bridgeNodes.Add(cluster2Head); // Agregar líder del clúster 2
    bridgeNodes.Add(bridgeCluster.Get(0)); // Agregar nodo puente
    ConfigureCluster(bridgeCluster.Get(0), bridgeNodes, 100.0); // Configura el puente

    // Configurar tráfico de tipo CBR
    OnOffHelper onOff("ns3::UdpSocketFactory", InetSocketAddress(Ipv4Address("10.1.0.255"), 9)); // CBR con tráfico UDP
    onOff.SetConstantRate(DataRate("500kbps")); // Configura la tasa constante de datos
    ApplicationContainer apps = onOff.Install(cluster1.Get(0)); // Instala la aplicación en el nodo 0 del clúster 1
    apps.Start(Seconds(1.0)); // Inicia el tráfico en el segundo 1
    apps.Stop(Seconds(10.0)); // Finaliza el tráfico en el segundo 10

    // Configurar tráfico FTP (TCP)
    BulkSendHelper bulkSend("ns3::TcpSocketFactory", InetSocketAddress(Ipv4Address("10.1.1.1"), 20)); // Tráfico FTP
    bulkSend.SetAttribute("MaxBytes", UintegerValue(1024 * 1024)); // Límite de 1 MB
    apps = bulkSend.Install(cluster2.Get(0)); // Instala la aplicación en el nodo 0 del clúster 2
    apps.Start(Seconds(1.0)); // Inicia el tráfico en el segundo 1
    apps.Stop(Seconds(10.0)); // Finaliza el tráfico en el segundo 10

    // Configurar FlowMonitor para análisis de rendimiento
    FlowMonitorHelper flowmon;
    Ptr<FlowMonitor> monitor = flowmon.InstallAll(); // Instala FlowMonitor en todos los nodos

    Simulator::Stop(Seconds(10.0)); // Configura la duración de la simulación
    Simulator::Run(); // Ejecuta la simulación

    // Guardar resultados en un archivo XML
    monitor->SerializeToXmlFile("manet-simulation.xml", true, true); // Exporta métricas a un archivo

    Simulator::Destroy(); // Limpia los recursos después de la simulación
    return 0;
}

File: hierarchical_manet.py
import ns.applications as apps
import ns.core as core
import ns.internet as internet
import ns.mobility as mobility
import ns.network as network
import ns.wifi as wifi
import ns.flow_monitor as flowmon

# Habilitar el registro para depuración
core.LogComponentEnable("HierarchicalMANET", core.LOG_LEVEL_INFO)

# Crear nodos para los clústeres
cluster1 = network.NodeContainer()
cluster2 = network.NodeContainer()
bridge_cluster = network.NodeContainer()

# Crear nodos en los clústeres
cluster1.Create(5)  # Clúster 1 con 5 nodos
cluster2.Create(5)  # Clúster 2 con 5 nodos
bridge_cluster.Create(1)  # Nodo puente para conectar clústeres

# Crear nodos líderes (cluster heads)
cluster1_head = network.Node()
cluster2_head = network.Node()

# Crear contenedor para puente entre clústeres
bridge_nodes = network.NodeContainer()
bridge_nodes.Add(cluster1_head)
bridge_nodes.Add(cluster2_head)
bridge_nodes.Add(bridge_cluster.Get(0))

# Función para configurar un clúster
def configure_cluster(cluster_head, cluster_nodes, range):
    """
    Configura un clúster en la red MANET.
    :param cluster_head: Nodo principal (líder del clúster)
    :param cluster_nodes: Nodos del clúster
    :param range: Rango de movilidad y posiciones
    """
    # Configurar dispositivos Wi-Fi
    wifi_helper = wifi.WifiHelper()
    wifi_helper.SetStandard(wifi.WIFI_PHY_STANDARD_80211g)
    
    mac = wifi.WifiMacHelper()
    mac.SetType("ns3::AdhocWifiMac")  # Configuración para modo ad hoc
    
    phy = wifi.YansWifiPhyHelper.Default()
    channel = wifi.YansWifiChannelHelper.Default()
    phy.SetChannel(channel.Create())
    
    # Instalar dispositivos Wi-Fi en los nodos
    devices = wifi_helper.Install(phy, mac, cluster_nodes)
    devices.Add(wifi_helper.Install(phy, mac, cluster_head))
    
    # Configurar pila de protocolos de Internet
    internet_stack = internet.InternetStackHelper()
    internet_stack.Install(cluster_nodes)
    internet_stack.Install(cluster_head)
    
    # Asignar direcciones IP
    ipv4 = internet.Ipv4AddressHelper()
    ipv4.SetBase("10.{}.0.0".format(range), "255.255.255.0")
    ipv4.Assign(devices)
    
    # Configurar movilidad de los nodos
    mobility_helper = mobility.MobilityHelper()
    position_alloc = mobility.ListPositionAllocator()
    
    # Asignar posiciones para los nodos
    for i in range(cluster_nodes.GetN()):
        position_alloc.Add(core.Vector3D(range + i * 10.0, range, 0.0))
    position_alloc.Add(core.Vector3D(range, range, 0.0))  # Nodo principal
    
    mobility_helper.SetPositionAllocator(position_alloc)
    mobility_helper.SetMobilityModel("ns3::RandomWalk2dMobilityModel",
                                     "Bounds", core.RectangleValue(core.Rectangle(-range, range, -range, range)))
    mobility_helper.Install(cluster_nodes)
    mobility_helper.Install(cluster_head)

# Configurar los tres clústeres
configure_cluster(cluster1_head, cluster1, 50)  # Clúster 1
configure_cluster(cluster2_head, cluster2, 150)  # Clúster 2
configure_cluster(bridge_cluster.Get(0), bridge_nodes, 100)  # Nodo puente

# Configurar tráfico CBR (Constant Bit Rate)
on_off_helper = apps.OnOffHelper("ns3::UdpSocketFactory", 
                                 internet.Ipv4AddressHelper.GetAddress("10.50.0.1", 9))
on_off_helper.SetConstantRate(core.DataRate("500kbps"))  # Tasa constante de datos
on_off_apps = on_off_helper.Install(cluster1.Get(0))  # Instalar en el primer nodo del clúster 1
on_off_apps.Start(core.Seconds(1.0))  # Inicio del tráfico
on_off_apps.Stop(core.Seconds(10.0))  # Fin del tráfico

# Configurar tráfico FTP
ftp_helper = apps.BulkSendHelper("ns3::TcpSocketFactory", 
                                 internet.Ipv4AddressHelper.GetAddress("10.150.0.1", 20))
ftp_helper.SetAttribute("MaxBytes", core.UintegerValue(1024 * 1024))  # Limitar a 1 MB
ftp_apps = ftp_helper.Install(cluster2.Get(0))  # Instalar en el primer nodo del clúster 2
ftp_apps.Start(core.Seconds(1.0))  # Inicio del tráfico
ftp_apps.Stop(core.Seconds(10.0))  # Fin del tráfico

# Configurar FlowMonitor para analizar resultados
flow_monitor_helper = flowmon.FlowMonitorHelper()
monitor = flow_monitor_helper.InstallAll()

# Ejecutar simulación
core.Simulator.Stop(core.Seconds(10.0))
core.Simulator.Run()

# Analizar resultados con FlowMonitor
monitor.SerializeToXmlFile("manet_hierarchical_simulation.xml", True, True)

core.Simulator.Destroy()
